### PR TITLE
add sample for chunking for upload using one drive connector

### PR DESCRIPTION
**Pull request by jspa**
add sample for chunking for upload using one drive connector

### CODE REVIEW

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+using Microsoft.Extensions.Logging; 
+using System.Collections.Generic;
+using Microsoft.Rest;
+using System.Threading;
+
+// *******One-Time Setup*********
+// Add Azure GitHub Organization as a source to nuget.config by following instructions below:
+// Follow instructions here to genereate GitHub personal access token with "read:packages" permissions
+// https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token
+// Once you have the PAT handy, run command "dotnet nuget add source https://nuget.pkg.github.com/Azure/index.json --name AzureGPR --username --password <PAT> --store-password-in-clear-text"
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs]: update the comment to include username

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            // get OneDrive connection string from app setting and set up connector
+            var oneDriveConnectionString = System.Environment.GetEnvironmentVariable("ONEDRIVE_CONNECTION", EnvironmentVariableTarget.Process);
+            var oneDriveConnector = OneDriveConnector.Create(oneDriveConnectionString);
+
+            // get reference to large file and download it, for download in chunks see https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-handle-large-messages#download-content-in-chunks for more info
+            var bigFileMetadata = await oneDriveConnector.OneDriveFileData.GetFileMetadataByPathAsync("Documents/largefile.zip");
+            var bigFile = await oneDriveConnector.OneDriveFileData.GetFileContentAsync(bigFileMetadata.Id);
+
+            // make initial call to find chunk size, see https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-handle-large-messages#upload-content-in-chunks for info
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs:71]: create a seperate method for getting the chunking recommendation

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        public static async Task<IActionResult> Run(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req,
+            ILogger log)
+        {
+            // get OneDrive connection string from app setting and set up connector
+            var oneDriveConnectionString = System.Environment.GetEnvironmentVariable("ONEDRIVE_CONNECTION", EnvironmentVariableTarget.Process);
+            var oneDriveConnector = OneDriveConnector.Create(oneDriveConnectionString);
+
+            // get reference to large file and download it, for download in chunks see https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-handle-large-messages#download-content-in-chunks for more info
+            var bigFileMetadata = await oneDriveConnector.OneDriveFileData.GetFileMetadataByPathAsync("Documents/largefile.zip");
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs:42]: The expectation to run this sample is there exists a file in users Documents.
Is it possible to upload a file before this step without chunking and then do the rest.
With the expectation the sample just works without the manual step.

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            log.LogInformation("C# HTTP trigger function processed a request.");
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
+
+            string responseMessage =  "This HTTP triggered function executed successfully.";
+
+            return new OkObjectResult(responseMessage);
+        }
+
+        private static async Task ConvertToChunks(Stream fileStream, int fileSize, int chunkSize, HttpClient httpClient, ServiceClientCredentials credentials, string location)
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs:95]: Not all connectors support chunking. How do we know the operation allows chunking. Are there comments?

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            var location = response.Response.Headers.GetValues("Location").First();
+
+            // depending on the upload speed, you may want to change the chunk size and not use the one suggested
+            chunkSize = "1000000";
+
+            await ConvertToChunks(bigFile, (int)bigFileMetadata.Size, int.Parse(chunkSize), oneDriveConnector.HttpClient, oneDriveConnector.Credentials, location);
+
+            log.LogInformation("C# HTTP trigger function processed a request.");
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs]: is this needed?

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                customHeaders: customHeaders);
+
+            // from response, find the recommended chunk size and location to make subsequent patch calls
+            var chunkSize = response.Response.Headers.GetValues("x-ms-chunk-size").First();
+            var location = response.Response.Headers.GetValues("Location").First();
+
+            // depending on the upload speed, you may want to change the chunk size and not use the one suggested
+            chunkSize = "1000000";
+
+            await ConvertToChunks(bigFile, (int)bigFileMetadata.Size, int.Parse(chunkSize), oneDriveConnector.HttpClient, oneDriveConnector.Credentials, location);
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs]: should this return to say the upload completed sucessfully

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+            log.LogInformation("C# HTTP trigger function processed a request.");
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
+
+            string responseMessage =  "This HTTP triggered function executed successfully.";
+
+            return new OkObjectResult(responseMessage);
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs]: return message if the upload has completed successfully

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                httpRequest.Content = new StreamContent(new MemoryStream(bytes));
+
+                httpRequest.Content.Headers.ContentRange = new System.Net.Http.Headers.ContentRangeHeaderValue(startIndex, endIndex-1, fileSize);
+                httpRequest.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/octet-stream");
+                httpRequest.Content.Headers.ContentLength = bytes.Length;
+
+                // set up credentials for the http request
+                await credentials.ProcessHttpRequestAsync(httpRequest, default(CancellationToken));
+
+                var response = await httpClient.SendAsync(httpRequest);
```
Comment (annaji-msft) [inline samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs]: try catch and return failed?

**Review by annaji-msft — APPROVED** [review]

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs (hunk 4 above)
Comment (jspa) [reply samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs:95]: added note in comments

File: samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs (hunk 3 above)
Comment (jspa) [reply samples/azurefunctions/backupFileOnOneDriveUsingChunking/csharp/BackupFileOnOneDriveUsingChunking.cs:42]: i would rather not do that.. cause i would need to have the functionapp itself accept a file as a request which would complicate the sample i think? I have added more comments in the readme to replace the file name to what they have in the onedrive.

**Pull request merged**